Meteor 1.8.0.1 with Angular 7.1.0 fixes - Fixes #1926

georgyberdyshev opened the pull request:
Hello @Urigo and @ardatan,

this PR has a couple of fixes for the MeteorCLI examples. Fixes #1926
Tested using Meteor 1.7.0.3 with Angular 6.0.7

**EDIT: Updated to Meteor 1.8.0.1 with Angular 7.1.0**

It would be great, if you could get this merged.

**Changeset**
- Add missing rxjs imports for filter, map and mergeMap.
Fixes error: Property '...' does not exist on type 'Observable<Event>'.
- Fixed Angular Lazy Loading of TodoAddModule.
- Updated to Meteor 1.6.1.3 release.
- Updated to Angular 5.2.11.
  Updated packages.
- Updated to Meteor 1.7.0.3 release.
- Updated to Angular 6.0.7.
  Updated packages.
- Updated Angular Meteor compiler packages to 0.3.2
- Convert to RxJS 6 Operator Pipe Syntax.
-Added Meteor, Node.js and missing package types to TypeScript compiler options.
 Fixes error: Cannot find name 'require'.
 Fixes error: Cannot find name 'Assets'.
- Fixed TypeScript target to use es2015.


Thanks, Georgy

Urigo commented:
thank you very much @georgyberdyshev !
just from first glance, could you check and see why the CI is failing?

georgyberdyshev commented:
Hi @Urigo,

thank you for your reply.

I will have a look and fix the CI error.

Georgy

ardatan commented:
@georgyberdyshev Thank you very much! 
Did you see my comment on the lazy loading part of the `app.module.ts`?

georgyberdyshev commented:
Hi @ardatan,
could you please let me know, which comment you refer to.
The lazy loading app.module.ts does not have any comments (client/imports/app/app.module.ts).

@Urigo and @ardatan I have fixed some issues with the CI and will open a PR for ardatan:mocha.
There were some missing asset issues.

Thanks, Georgy

ardatan commented on examples/MeteorCLI/all-in-one/imports/app/app.module.ts:
```diff
@@ -27,7 +28,7 @@ import { PageNotFoundComponent } from './page-not-found/page-not-found.component
       },
       {
         path: 'todoAdd',
-        loadChildren: './todo-add/todo-add.module#TodoAddModule',
```
This line is replaced by compiler with Meteor's dynamic loading syntax, so if you add it regular way, this isn't going to be loaded lazily.

georgyberdyshev replied:
Thanks @ardatan!
I must have missed it. There were some initial errors, while still upgrading to a newer Angular Meteor compiler version.

ardatan commented:
@georgyberdyshev This line is replaced by compiler with Meteor's dynamic loading syntax, so if you add it regular way, this isn't going to be loaded lazily.

georgyberdyshev commented:
@ardatan thanks for merging my PR for meteor-mocha and for the comments regarding lazy loading.

I must have missed the lazy loading part.  When migrating to Angular 6, there were still some errors due to a not-updated Angular Meteor compiler version.

Reverted it and now it is working.

Thanks, Georgy

ardatan commented:
@georgyberdyshev Thank you! Somehow, mocha gives some errors like below only in Travis. I couldn't find a solution for then. Do you have any idea?
```
TypeError: undefined is not an object (evaluating 'Package.xxx)
```

georgyberdyshev commented:
Hello @ardatan,

the Travis CI errors require a specific configuration to reproduce.

**On Travis CI, there is a PhantomJS binary already installed:**
```
Considering PhantomJS found at /usr/local/phantomjs/bin/phantomjs
Found PhantomJS at /usr/local/phantomjs/bin/phantomjs ...verifying
Writing location.js file
PhantomJS is already installed on PATH at /usr/local/phantomjs/bin/phantomjs
```

**To reproduce the same error, you have to make sure that PhantomJS is not installed
and is not set in your $PATH:**

I could reproduce it on Ubuntu 18.04, but it should also work on other Linux distributions
as long as you do not have PhantomJS.

To make sure, just check:
```
which phantomjs
phantomjs not found
```

In **angular-meteor/examples/MeteorCLI/bare** run **meteor npm install**:

This will download the correct prebuilt version of PhantomJS:
```
meteor npm install 

PhantomJS not found on PATH
Downloading https://github.com/Medium/phantomjs/releases/download/v2.1.1/phantomjs-2.1.1-linux-x86_64.tar.bz2
Saving to /tmp/phantomjs/phantomjs-2.1.1-linux-x86_64.tar.bz2
Receiving...
  [=====================================---] 94%
Received 22866K total.
Extracting tar contents (via spawned process)

Done. Phantomjs binary available at angular-meteor/examples/MeteorCLI/bare/node_modules/phantomjs-prebuilt/lib/phantom/bin/phantomjs
```

**After that, you can run the tests that will show the same errors as on Travis CI:**
`npm run test:ci`


**Now to the caveat:**
If you have PhantomJS installed you will get another type of error.

```
sudo apt install phantomjs
which phantomjs
/usr/bin/phantomjs

phantomjs --version
2.1.1
```


The error looks like this:
```
npm run test:ci

> @ test:ci angular-meteor/examples/MeteorCLI/bare
> TEST_BROWSER_DRIVER=phantomjs meteor test --driver-package=ardatan:mocha --once

[[[[[ Tests ]]]]]                             

=> Started proxy.                             
...
...
=> Started MongoDB.                           
?                    
? --------------------------------
? ----- RUNNING SERVER TESTS -----
? --------------------------------
? 
? 
? 
?   0 passing (1ms)
? 
? 
? --------------------------------
? ----- RUNNING CLIENT TESTS -----
? --------------------------------
=> Started your app.

=> App running at: http://localhost:3000/
? QStandardPaths: XDG_RUNTIME_DIR not set, defaulting to '/tmp/runtime-asdf'
? qt.qpa.screen: QXcbConnection: Could not connect to display 
? Could not connect to any X display.
? All tests finished!
? 
? --------------------------------
? SERVER FAILURES: 0
? CLIENT FAILURES: 1
? --------------------------------
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! @ test:ci: `TEST_BROWSER_DRIVER=phantomjs meteor test --driver-package=ardatan:mocha --once`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the @ test:ci script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     .npm/_logs/asdf-debug.log

```


This can be **fixed by setting the following environment variable**, which enables headless rendering, for the local PhantomJS installation:

```
QT_QPA_PLATFORM=offscreen phantomjs --version
2.1.1
```


Although this workaround solves the issues when PhantomJS was installed from a package, it causes an error when used with the prebuilt version.

@ardatan, I will send a PR for ardatan:mocha that checks for a local installation of PhantomJS and otherwise falls back to the prebuilt version.
But we will still need to figure out what is causing the errors of the prebuilt version.

**As you can see below, the tests are now working:**

```
npm run test:ci

> @ test:ci angular-meteor/examples/MeteorCLI/bare
> TEST_BROWSER_DRIVER=phantomjs meteor test --driver-package=ardatan:mocha --once

[[[[[ Tests ]]]]]                             

=> Started proxy.                             
...
...
=> Started MongoDB.                           
?                    
? --------------------------------
? ----- RUNNING SERVER TESTS -----
? --------------------------------
? 
? 
? 
?   0 passing (1ms)
? 
? 
? --------------------------------
? ----- RUNNING CLIENT TESTS -----
? --------------------------------
=> Started your app.

=> App running at: http://localhost:3000/
? 
? 
?   TodoListComponent
?     1) should display `a` element
?     2) should render styles correctly
? 
?   AppComponent
?     3) should display h1 element
? 
? 
?   0 passing (446ms)
?   3 failing
? 
?   1) TodoListComponent
?        should display `a` element:
?      chai_1.expect(a.nativeElement.textContent).to.equal is not a function. (In 'chai_1.expect(a.nativeElement.textContent).to.equal('Add')', 'chai_1.expect(a.nativeElement.textContent).to.equal' is an instance of ProxyObject)--Pendng async tasks are: [type: microTask, source: Promise.then, args: {__zone_symbol__state:null,__zone_symbol__value:[object Object],[object Promise],,function (e) { _this.reject(e); }}]
?   http://localhost:3000/app.js?hash=d225493391b3199f56e3aef9a286e514b6ed21f4:30:60
?   run@http://localhost:3000/packages/modules.js?hash=5bcd09233ec2b37a89ade2f1ef1af85e0b63f7c4:1372:49
?   callFn@http://localhost:3000/packages/modules.js?hash=5bcd09233ec2b37a89ade2f1ef1af85e0b63f7c4:108876:25
?   run@http://localhost:3000/packages/modules.js?hash=5bcd09233ec2b37a89ade2f1ef1af85e0b63f7c4:108868:13
?   runTest@http://localhost:3000/packages/modules.js?hash=5bcd09233ec2b37a89ade2f1ef1af85e0b63f7c4:109404:13
?   http://localhost:3000/packages/modules.js?hash=5bcd09233ec2b37a89ade2f1ef1af85e0b63f7c4:5817:37
?   runTask@http://localhost:3000/packages/modules.js?hash=5bcd09233ec2b37a89ade2f1ef1af85e0b63f7c4:1422:57
?   drainMicroTaskQueue@http://localhost:3000/packages/modules.js?hash=5bcd09233ec2b37a89ade2f1ef1af85e0b63f7c4:1829:42
?   invokeTask@http://localhost:3000/packages/modules.js?hash=5bcd09233ec2b37a89ade2f1ef1af85e0b63f7c4:1734:40
?   timer@http://localhost:3000/packages/modules.js?hash=5bcd09233ec2b37a89ade2f1ef1af85e0b63f7c4:3288:34
? 
?   2) TodoListComponent
?        should render styles correctly:
?      chai_1.expect(color).to.equal is not a function. (In 'chai_1.expect(color).to.equal('rgb(0, 255, 0)')', 'chai_1.expect(color).to.equal' is an instance of ProxyObject)--Pendng async tasks are: [type: microTask, source: Promise.then, args: {__zone_symbol__state:null,__zone_symbol__value:[object Object],[object Promise],,function (e) { _this.reject(e); }}]
?   http://localhost:3000/app.js?hash=d225493391b3199f56e3aef9a286e514b6ed21f4:37:38
?   run@http://localhost:3000/packages/modules.js?hash=5bcd09233ec2b37a89ade2f1ef1af85e0b63f7c4:1372:49
?   callFn@http://localhost:3000/packages/modules.js?hash=5bcd09233ec2b37a89ade2f1ef1af85e0b63f7c4:108876:25
?   run@http://localhost:3000/packages/modules.js?hash=5bcd09233ec2b37a89ade2f1ef1af85e0b63f7c4:108868:13
?   runTest@http://localhost:3000/packages/modules.js?hash=5bcd09233ec2b37a89ade2f1ef1af85e0b63f7c4:109404:13
?   http://localhost:3000/packages/modules.js?hash=5bcd09233ec2b37a89ade2f1ef1af85e0b63f7c4:5817:37
?   runTask@http://localhost:3000/packages/modules.js?hash=5bcd09233ec2b37a89ade2f1ef1af85e0b63f7c4:1422:57
?   drainMicroTaskQueue@http://localhost:3000/packages/modules.js?hash=5bcd09233ec2b37a89ade2f1ef1af85e0b63f7c4:1829:42
?   invokeTask@http://localhost:3000/packages/modules.js?hash=5bcd09233ec2b37a89ade2f1ef1af85e0b63f7c4:1734:40
?   timer@http://localhost:3000/packages/modules.js?hash=5bcd09233ec2b37a89ade2f1ef1af85e0b63f7c4:3288:34
? 
?   3) AppComponent
?        should display h1 element:
?      chai_1.expect(h1.nativeElement.textContent).to.equal is not a function. (In 'chai_1.expect(h1.nativeElement.textContent).to.equal('Todos')', 'chai_1.expect(h1.nativeElement.textContent).to.equal' is an instance of ProxyObject)--Pendng async tasks are: [type: microTask, source: Promise.then, args: {__zone_symbol__state:null,__zone_symbol__value:[object Object],[object Promise],,function (e) { _this.reject(e); }}]
?   http://localhost:3000/app.js?hash=d225493391b3199f56e3aef9a286e514b6ed21f4:122:61
?   run@http://localhost:3000/packages/modules.js?hash=5bcd09233ec2b37a89ade2f1ef1af85e0b63f7c4:1372:49
?   callFn@http://localhost:3000/packages/modules.js?hash=5bcd09233ec2b37a89ade2f1ef1af85e0b63f7c4:108876:25
?   run@http://localhost:3000/packages/modules.js?hash=5bcd09233ec2b37a89ade2f1ef1af85e0b63f7c4:108868:13
?   runTest@http://localhost:3000/packages/modules.js?hash=5bcd09233ec2b37a89ade2f1ef1af85e0b63f7c4:109404:13
?   http://localhost:3000/packages/modules.js?hash=5bcd09233ec2b37a89ade2f1ef1af85e0b63f7c4:5817:37
?   runTask@http://localhost:3000/packages/modules.js?hash=5bcd09233ec2b37a89ade2f1ef1af85e0b63f7c4:1422:57
?   drainMicroTaskQueue@http://localhost:3000/packages/modules.js?hash=5bcd09233ec2b37a89ade2f1ef1af85e0b63f7c4:1829:42
?   invokeTask@http://localhost:3000/packages/modules.js?hash=5bcd09233ec2b37a89ade2f1ef1af85e0b63f7c4:1734:40
?   timer@http://localhost:3000/packages/modules.js?hash=5bcd09233ec2b37a89ade2f1ef1af85e0b63f7c4:3288:34
? 
? 
? 
? TypeError: Attempting to change the setter of an unconfigurable property.
? TypeError: Attempting to change the setter of an unconfigurable property.
? All tests finished!
? 
? --------------------------------
? SERVER FAILURES: 0
? CLIENT FAILURES: 3
? --------------------------------
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! @ test:ci: `TEST_BROWSER_DRIVER=phantomjs meteor test --driver-package=ardatan:mocha --once`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the @ test:ci script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     .npm/_logs/asdf-debug.log
```

Thanks, Georgy

ardatan commented:
@georgyberdyshev Unfortunately, the error still occurs. Maybe, we should try https://github.com/meteortesting/meteor-mocha

georgyberdyshev commented:
Hello @ardatan,

just added an additional PATH configuration for Travis CI in https://github.com/ardatan/meteor-mocha/pull/5
Let's wait for Travis CI and see.

I did test meteortesting:mocha and the error is the same:

```
TEST_BROWSER_DRIVER=phantomjs meteor test --driver-package=meteortesting:mocha --once
...
TypeError: undefined is not an object (evaluating 'Package.xxx)
...
```

Georgy

georgyberdyshev commented:
@ardatan it now breaks due to QT_QPA_PLATFORM=offscreen

On Travis CI:
`This application failed to start because it could not find or load the Qt platform plugin "offscreen".`

It is still required. Otherwise you get:
`=> App running at: http://localhost:3000/
? QStandardPaths: XDG_RUNTIME_DIR not set, defaulting to '/tmp/runtime-asdf'
? qt.qpa.screen: QXcbConnection: Could not connect to display 
? Could not connect to any X display.
? All tests finished!`

So, I will have to add some changes testing also for this in regard to Travis CI.

Georgy

georgyberdyshev commented:
Hello @ardatan,

I added a new PR, fixing the offscreen plugin issue: https://github.com/ardatan/meteor-mocha/pull/6

Thanks, Georgy

ardatan commented:
Sorry, but the error still exists.

georgyberdyshev commented:
Hello @ardatan,

the error is related to the PhantomJS installation on Travis CI.

Their environment is based on Ubuntu 14.04 (Trusty).

The default apt package installation would install to: /usr/bin/phantomjs
Travis CI has a custom version in: /usr/local/phantomjs/bin

You could try to change .travis.yml and install PhantonJS 1.9.0 for Ubuntu 14.04
https://packages.ubuntu.com/search?keywords=phantomjs&searchon=names&suite=trusty&section=all

Then we can try to switch between the package installation and Travis CI custom installation.
This would be very helpful!

If you check on a Ubuntu 18.04 desktop installation, you can see that the tests are running and the output is:

```
npm run test:ci

> @ test:ci angular-meteor/examples/MeteorCLI/bare
> TEST_BROWSER_DRIVER=phantomjs meteor test --driver-package=ardatan:mocha --once

[[[[[ Tests ]]]]]                             

=> Started proxy.                             
...
...
=> Started MongoDB.                           
?                    
? --------------------------------
? ----- RUNNING SERVER TESTS -----
? --------------------------------
? 
? 
? 
?   0 passing (1ms)
? 
? 
? --------------------------------
? ----- RUNNING CLIENT TESTS -----
? --------------------------------
? Checking offscreen plugin support.
? Offscreen plugin found!
=> Started your app.

=> App running at: http://localhost:3000/
? 
? 
?   TodoListComponent
?     1) should display `a` element
?     2) should render styles correctly
? 
?   AppComponent
?     3) should display h1 element
? 
? 
?   0 passing (539ms)
?   3 failing
? 
?   1) TodoListComponent
?        should display `a` element:
?      chai_1.expect(a.nativeElement.textContent).to.equal is not a function. (In 'chai_1.expect(a.nativeElement.textContent).to.equal('Add')', 'chai_1.expect(a.nativeElement.textContent).to.equal' is an instance of ProxyObject)--Pendng async tasks are: [type: microTask, source: Promise.then, args: {__zone_symbol__state:null,__zone_symbol__value:[object Object],[object Promise],,function (e) { _this.reject(e); }}]
?   http://localhost:3000/app.js?hash=d225493391b3199f56e3aef9a286e514b6ed21f4:30:60
?   run@http://localhost:3000/packages/modules.js?hash=5bcd09233ec2b37a89ade2f1ef1af85e0b63f7c4:1372:49
?   callFn@http://localhost:3000/packages/modules.js?hash=5bcd09233ec2b37a89ade2f1ef1af85e0b63f7c4:108876:25
?   run@http://localhost:3000/packages/modules.js?hash=5bcd09233ec2b37a89ade2f1ef1af85e0b63f7c4:108868:13
?   runTest@http://localhost:3000/packages/modules.js?hash=5bcd09233ec2b37a89ade2f1ef1af85e0b63f7c4:109404:13
?   http://localhost:3000/packages/modules.js?hash=5bcd09233ec2b37a89ade2f1ef1af85e0b63f7c4:5817:37
?   runTask@http://localhost:3000/packages/modules.js?hash=5bcd09233ec2b37a89ade2f1ef1af85e0b63f7c4:1422:57
?   drainMicroTaskQueue@http://localhost:3000/packages/modules.js?hash=5bcd09233ec2b37a89ade2f1ef1af85e0b63f7c4:1829:42
?   invokeTask@http://localhost:3000/packages/modules.js?hash=5bcd09233ec2b37a89ade2f1ef1af85e0b63f7c4:1734:40
?   timer@http://localhost:3000/packages/modules.js?hash=5bcd09233ec2b37a89ade2f1ef1af85e0b63f7c4:3288:34
? 
?   2) TodoListComponent
?        should render styles correctly:
?      chai_1.expect(color).to.equal is not a function. (In 'chai_1.expect(color).to.equal('rgb(0, 255, 0)')', 'chai_1.expect(color).to.equal' is an instance of ProxyObject)--Pendng async tasks are: [type: microTask, source: Promise.then, args: {__zone_symbol__state:null,__zone_symbol__value:[object Object],[object Promise],,function (e) { _this.reject(e); }}]
?   http://localhost:3000/app.js?hash=d225493391b3199f56e3aef9a286e514b6ed21f4:37:38
?   run@http://localhost:3000/packages/modules.js?hash=5bcd09233ec2b37a89ade2f1ef1af85e0b63f7c4:1372:49
?   callFn@http://localhost:3000/packages/modules.js?hash=5bcd09233ec2b37a89ade2f1ef1af85e0b63f7c4:108876:25
?   run@http://localhost:3000/packages/modules.js?hash=5bcd09233ec2b37a89ade2f1ef1af85e0b63f7c4:108868:13
?   runTest@http://localhost:3000/packages/modules.js?hash=5bcd09233ec2b37a89ade2f1ef1af85e0b63f7c4:109404:13
?   http://localhost:3000/packages/modules.js?hash=5bcd09233ec2b37a89ade2f1ef1af85e0b63f7c4:5817:37
?   runTask@http://localhost:3000/packages/modules.js?hash=5bcd09233ec2b37a89ade2f1ef1af85e0b63f7c4:1422:57
?   drainMicroTaskQueue@http://localhost:3000/packages/modules.js?hash=5bcd09233ec2b37a89ade2f1ef1af85e0b63f7c4:1829:42
?   invokeTask@http://localhost:3000/packages/modules.js?hash=5bcd09233ec2b37a89ade2f1ef1af85e0b63f7c4:1734:40
?   timer@http://localhost:3000/packages/modules.js?hash=5bcd09233ec2b37a89ade2f1ef1af85e0b63f7c4:3288:34
? 
?   3) AppComponent
?        should display h1 element:
?      chai_1.expect(h1.nativeElement.textContent).to.equal is not a function. (In 'chai_1.expect(h1.nativeElement.textContent).to.equal('Todos')', 'chai_1.expect(h1.nativeElement.textContent).to.equal' is an instance of ProxyObject)--Pendng async tasks are: [type: microTask, source: Promise.then, args: {__zone_symbol__state:null,__zone_symbol__value:[object Object],[object Promise],,function (e) { _this.reject(e); }}]
?   http://localhost:3000/app.js?hash=d225493391b3199f56e3aef9a286e514b6ed21f4:122:61
?   run@http://localhost:3000/packages/modules.js?hash=5bcd09233ec2b37a89ade2f1ef1af85e0b63f7c4:1372:49
?   callFn@http://localhost:3000/packages/modules.js?hash=5bcd09233ec2b37a89ade2f1ef1af85e0b63f7c4:108876:25
?   run@http://localhost:3000/packages/modules.js?hash=5bcd09233ec2b37a89ade2f1ef1af85e0b63f7c4:108868:13
?   runTest@http://localhost:3000/packages/modules.js?hash=5bcd09233ec2b37a89ade2f1ef1af85e0b63f7c4:109404:13
?   http://localhost:3000/packages/modules.js?hash=5bcd09233ec2b37a89ade2f1ef1af85e0b63f7c4:5817:37
?   runTask@http://localhost:3000/packages/modules.js?hash=5bcd09233ec2b37a89ade2f1ef1af85e0b63f7c4:1422:57
?   drainMicroTaskQueue@http://localhost:3000/packages/modules.js?hash=5bcd09233ec2b37a89ade2f1ef1af85e0b63f7c4:1829:42
?   invokeTask@http://localhost:3000/packages/modules.js?hash=5bcd09233ec2b37a89ade2f1ef1af85e0b63f7c4:1734:40
?   timer@http://localhost:3000/packages/modules.js?hash=5bcd09233ec2b37a89ade2f1ef1af85e0b63f7c4:3288:34
? 
? 
? 
? TypeError: Attempting to change the setter of an unconfigurable property.
? TypeError: Attempting to change the setter of an unconfigurable property.
? All tests finished!
? 
? --------------------------------
? SERVER FAILURES: 0
? CLIENT FAILURES: 3
? --------------------------------
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! @ test:ci: `TEST_BROWSER_DRIVER=phantomjs meteor test --driver-package=ardatan:mocha --once`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the @ test:ci script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     .npm/_logs/asdf-debug.log
```

Thanks, Georgy

georgyberdyshev commented:
@ardatan could you please check #1929?

ardatan commented:
@georgyberdyshev Why did you seperate those changes? I think it'd better to do that change in this PR?

ardatan commented:
Are you sure we have a good point like qt, phantomjs binary things etc? Because the errors have been never changed?

georgyberdyshev commented:
@ardatan this PR https://github.com/ardatan/meteor-mocha/pull/7 changes the path order.

So that the Travis CI phantomjs path is overwritten when there is system package phantomjs installation.

ardatan commented:
@georgyberdyshev Failed again

georgyberdyshev commented:
@ardatan 

Seems to only work correctly with offscreen enabled, see the test output above: https://github.com/Urigo/angular-meteor/pull/1927#issuecomment-399793547

I can now reproduce the Travis CI error on Ubuntu 14.04

There are some issues in general with the package build: 
https://bugs.launchpad.net/ubuntu/+source/phantomjs/+bug/1586134
https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=817277

Thanks, Georgy

georgyberdyshev commented:
Hello @ardatan and @Urigo,

what do you think about using a Docker container on Travis CI that runs Ubuntu 18.04?
This will give us a working PhantonmJS version with Qt.

https://docs.travis-ci.com/user/docker/

Georgy

Urigo commented:
@georgyberdyshev sounds good to me

georgyberdyshev commented:
Hello @ardatan and @Urigo,

I have added a docker container running Ubuntu 18.04.

Node Version Manager (nvm) is used to install an LTS version of Node.js,
npm is updated to latest and also Meteor installed.

There is an issue with a sudden stop while installing npm dependencies:
-> [MeteorCLI - bare] Installing npm dependencies

You can check the raw logs of Travis CI.

Thanks, Georgy

ardatan commented:
@georgyberdyshev Thanks for your effort.
Could you fix this issue?

georgyberdyshev commented:
Hello @ardatan,

I am looking forward to getting this fixed and merged :+1: 

Next would be to verify, if it is a docker or a docker running inside Travis CI issue. Specifically why it stops at the npm dependencies installation.

After some initial research, Google-fu bug reports, it seems to be an out of memory error.

Georgy

georgyberdyshev commented:
Hello @ardatan and @Urigo,

the tests are running now running in the docker container, as you can see in the TravisCI build log.

I have also removed the rxjs-compat dependency and fixed the outstanding issues with rxjs.

Georgy

ardatan commented:
Thank you for everything.
But the tests need to be fixed before we merge it.

georgyberdyshev commented:
Hello @ardatan and @Urigo,

do you have any ideas on fixing the chai "... is not a function" errors?

I have tried several ways to import chai and / or expect directly, checked github issues and the documentation at http://www.chaijs.com/guide/styles/#expect

The error is:
```
[MeteorCLI - bare] Testing JIT
> @ test:ci /angular-meteor-sourcecode/examples/MeteorCLI/bare
> TEST_BROWSER_DRIVER=phantomjs meteor test --driver-package=ardatan:mocha --once
[[[[[ Tests ]]]]]
=> Started proxy.
[client]: Collecting TypeScript source files: 6.304ms
[client]: TypeScript Files Compilation: 6027.481ms
[client]: HTML Files Compilation: 9.110ms
[client]: SCSS Files Compilation: 3.463ms
[client]: Collecting TypeScript source files: 5.860ms
[client]: TypeScript Files Compilation: 2700.682ms
[client]: HTML Files Compilation: 3.273ms
[client]: SCSS Files Compilation: 1.393ms
[server]: Collecting TypeScript source files: 31.881ms
[server]: TypeScript Files Compilation: 1581.820ms
[server]: HTML Files Compilation: 0.017ms
=> Started MongoDB.
I20180712-09:11:23.177(0)? 
I20180712-09:11:23.230(0)? --------------------------------
I20180712-09:11:23.231(0)? ----- RUNNING SERVER TESTS -----
I20180712-09:11:23.232(0)? --------------------------------
I20180712-09:11:23.232(0)? 
I20180712-09:11:23.232(0)? 
I20180712-09:11:23.233(0)? 
I20180712-09:11:23.233(0)?   0 passing (0ms)
I20180712-09:11:23.234(0)? 
I20180712-09:11:23.234(0)? 
I20180712-09:11:23.235(0)? --------------------------------
I20180712-09:11:23.235(0)? ----- RUNNING CLIENT TESTS -----
I20180712-09:11:23.236(0)? --------------------------------
I20180712-09:11:23.236(0)? Checking offscreen plugin support.
I20180712-09:11:23.264(0)? Offscreen plugin found!
=> Started your app.
=> App running at: http://localhost:3000/
I20180712-09:11:24.451(0)? 
I20180712-09:11:24.452(0)? 
I20180712-09:11:24.457(0)?   TodoListComponent
I20180712-09:11:24.625(0)?     1) should display `a` element
I20180712-09:11:24.708(0)?     2) should render styles correctly
I20180712-09:11:24.720(0)? 
I20180712-09:11:24.721(0)?   AppComponent
I20180712-09:11:24.780(0)?     3) should display h1 element
I20180712-09:11:24.793(0)? 
I20180712-09:11:24.794(0)? 
I20180712-09:11:24.795(0)?   0 passing (344ms)
I20180712-09:11:24.795(0)?   3 failing
I20180712-09:11:24.796(0)? 
I20180712-09:11:24.796(0)?   1) TodoListComponent
I20180712-09:11:24.797(0)?        should display `a` element:
I20180712-09:11:24.797(0)?      chai_1.expect(a.nativeElement.textContent).to.equal is not a function. (In 'chai_1.expect(a.nativeElement.textContent).to.equal('Add')', 'chai_1.expect(a.nativeElement.textContent).to.equal' is an instance of ProxyObject)--Pendng async tasks are: [type: microTask, source: Promise.then, args: {__zone_symbol__state:null,__zone_symbol__value:[object Object],[object Promise],,function (e) { _this.reject(e); }}]
I20180712-09:11:24.797(0)?   http://localhost:3000/app.js?hash=d225493391b3199f56e3aef9a286e514b6ed21f4:30:60
I20180712-09:11:24.797(0)?   run@http://localhost:3000/packages/modules.js?hash=458cd9265212d82f90010000e285684af9321251:1372:49
I20180712-09:11:24.798(0)?   callFn@http://localhost:3000/packages/modules.js?hash=458cd9265212d82f90010000e285684af9321251:108441:25
I20180712-09:11:24.798(0)?   run@http://localhost:3000/packages/modules.js?hash=458cd9265212d82f90010000e285684af9321251:108433:13
I20180712-09:11:24.798(0)?   runTest@http://localhost:3000/packages/modules.js?hash=458cd9265212d82f90010000e285684af9321251:108969:13
I20180712-09:11:24.798(0)?   http://localhost:3000/packages/modules.js?hash=458cd9265212d82f90010000e285684af9321251:5817:37
I20180712-09:11:24.799(0)?   runTask@http://localhost:3000/packages/modules.js?hash=458cd9265212d82f90010000e285684af9321251:1422:57
I20180712-09:11:24.799(0)?   drainMicroTaskQueue@http://localhost:3000/packages/modules.js?hash=458cd9265212d82f90010000e285684af9321251:1829:42
I20180712-09:11:24.799(0)?   invokeTask@http://localhost:3000/packages/modules.js?hash=458cd9265212d82f90010000e285684af9321251:1734:40
I20180712-09:11:24.799(0)?   timer@http://localhost:3000/packages/modules.js?hash=458cd9265212d82f90010000e285684af9321251:3288:34
I20180712-09:11:24.801(0)? 
I20180712-09:11:24.801(0)?   2) TodoListComponent
I20180712-09:11:24.801(0)?        should render styles correctly:
I20180712-09:11:24.801(0)?      chai_1.expect(color).to.equal is not a function. (In 'chai_1.expect(color).to.equal('rgb(0, 255, 0)')', 'chai_1.expect(color).to.equal' is an instance of ProxyObject)--Pendng async tasks are: [type: microTask, source: Promise.then, args: {__zone_symbol__state:null,__zone_symbol__value:[object Object],[object Promise],,function (e) { _this.reject(e); }}]
I20180712-09:11:24.802(0)?   http://localhost:3000/app.js?hash=d225493391b3199f56e3aef9a286e514b6ed21f4:37:38
I20180712-09:11:24.802(0)?   run@http://localhost:3000/packages/modules.js?hash=458cd9265212d82f90010000e285684af9321251:1372:49
I20180712-09:11:24.802(0)?   callFn@http://localhost:3000/packages/modules.js?hash=458cd9265212d82f90010000e285684af9321251:108441:25
I20180712-09:11:24.802(0)?   run@http://localhost:3000/packages/modules.js?hash=458cd9265212d82f90010000e285684af9321251:108433:13
I20180712-09:11:24.803(0)?   runTest@http://localhost:3000/packages/modules.js?hash=458cd9265212d82f90010000e285684af9321251:108969:13
I20180712-09:11:24.803(0)?   http://localhost:3000/packages/modules.js?hash=458cd9265212d82f90010000e285684af9321251:5817:37
I20180712-09:11:24.803(0)?   runTask@http://localhost:3000/packages/modules.js?hash=458cd9265212d82f90010000e285684af9321251:1422:57
I20180712-09:11:24.803(0)?   drainMicroTaskQueue@http://localhost:3000/packages/modules.js?hash=458cd9265212d82f90010000e285684af9321251:1829:42
I20180712-09:11:24.804(0)?   invokeTask@http://localhost:3000/packages/modules.js?hash=458cd9265212d82f90010000e285684af9321251:1734:40
I20180712-09:11:24.804(0)?   timer@http://localhost:3000/packages/modules.js?hash=458cd9265212d82f90010000e285684af9321251:3288:34
I20180712-09:11:24.804(0)? 
I20180712-09:11:24.804(0)?   3) AppComponent
I20180712-09:11:24.805(0)?        should display h1 element:
I20180712-09:11:24.805(0)?      chai_1.expect(h1.nativeElement.textContent).to.equal is not a function. (In 'chai_1.expect(h1.nativeElement.textContent).to.equal('Todos')', 'chai_1.expect(h1.nativeElement.textContent).to.equal' is an instance of ProxyObject)--Pendng async tasks are: [type: microTask, source: Promise.then, args: {__zone_symbol__state:null,__zone_symbol__value:[object Object],[object Promise],,function (e) { _this.reject(e); }}]
I20180712-09:11:24.805(0)?   http://localhost:3000/app.js?hash=d225493391b3199f56e3aef9a286e514b6ed21f4:122:61
I20180712-09:11:24.805(0)?   run@http://localhost:3000/packages/modules.js?hash=458cd9265212d82f90010000e285684af9321251:1372:49
I20180712-09:11:24.806(0)?   callFn@http://localhost:3000/packages/modules.js?hash=458cd9265212d82f90010000e285684af9321251:108441:25
I20180712-09:11:24.806(0)?   run@http://localhost:3000/packages/modules.js?hash=458cd9265212d82f90010000e285684af9321251:108433:13
I20180712-09:11:24.806(0)?   runTest@http://localhost:3000/packages/modules.js?hash=458cd9265212d82f90010000e285684af9321251:108969:13
I20180712-09:11:24.806(0)?   http://localhost:3000/packages/modules.js?hash=458cd9265212d82f90010000e285684af9321251:5817:37
I20180712-09:11:24.807(0)?   runTask@http://localhost:3000/packages/modules.js?hash=458cd9265212d82f90010000e285684af9321251:1422:57
I20180712-09:11:24.807(0)?   drainMicroTaskQueue@http://localhost:3000/packages/modules.js?hash=458cd9265212d82f90010000e285684af9321251:1829:42
I20180712-09:11:24.807(0)?   invokeTask@http://localhost:3000/packages/modules.js?hash=458cd9265212d82f90010000e285684af9321251:1734:40
I20180712-09:11:24.807(0)?   timer@http://localhost:3000/packages/modules.js?hash=458cd9265212d82f90010000e285684af9321251:3288:34
I20180712-09:11:24.807(0)? 
I20180712-09:11:24.808(0)? 
I20180712-09:11:24.808(0)? 
I20180712-09:11:24.959(0)? TypeError: Attempting to change the setter of an unconfigurable property.
I20180712-09:11:24.959(0)? TypeError: Attempting to change the setter of an unconfigurable property.
I20180712-09:11:24.968(0)? All tests finished!
I20180712-09:11:24.969(0)? 
I20180712-09:11:24.969(0)? --------------------------------
I20180712-09:11:24.969(0)? SERVER FAILURES: 0
I20180712-09:11:24.970(0)? CLIENT FAILURES: 3
I20180712-09:11:24.970(0)? --------------------------------
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! @ test:ci: `TEST_BROWSER_DRIVER=phantomjs meteor test --driver-package=ardatan:mocha --once`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the @ test:ci script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
npm ERR! A complete log of this run can be found in:
npm ERR!     /home/docker/.npm/_logs/2018-07-12T09_11_25_243Z-debug.log
```

Thanks, Georgy

georgyberdyshev commented:
Hello @ardatan,

if you check the TravisCI build log for master, you will see that the tests were not working before.

Please let me know, if you have any ideas o the chai errors.

Thanks, Georgy

ardatan commented:
Could you replace ‘npm ci’ with ‘npm install’?

georgyberdyshev commented:
@ardatan Still the same error.

ardatan commented:
Let me checkout your branch and try it locally.

ardatan commented:
You can use git rebase to prevent a lot of commits for similar changes.

ardatan commented:
Hi @georgyberdyshev, could you replace all `npm ci` with `npm i`?

georgyberdyshev commented:
Hello @ardatan, replacing npm ci with npm i still produces the same test errors.

ardatan commented:
Thank you @georgyberdyshev !
You can use `v1.1.0` of `ardatan:meteor-mocha`

georgyberdyshev commented:
Hi @ardatan,

the tests for the MeteorCLI are working now.

But there is an issue with AOT w/ Rollup:
```
Compiling undefined with meteor-babel without a cache
[object Object]
...
...
```
Some good remarks regarding implementing setDiskCacheDirectory are in https://github.com/meteor/meteor/issues/10236#issuecomment-427048932

Furthermore, the version of @babel/runtime is outdated for the AngularCLI / angularjs and is
causing further errors.
Should be fixable with:
`meteor npm install --save @babel/runtime@latest`

Thanks, Georgy
